### DISCUSSION TITLE
using skforecast for missing value imputation

### DISCUSSION THREAD

**Discussion by slavakx** [JoaquinAmatRodrigo/skforecast#691] [General]
I have many time series with missing values of different size. Does skforecast suit the use case of training a regression model with time components like day, week, month but without explicitly generating lag features? The reason for not using lags is that I want to impute missing values for a quite long time series sequence. To be more general, are lag features required in all skforecast algorithms?

**Comment by JavierEscobarOrtiz** [comment]
Hello @slavakx 

Lag features are required in all Forecasters. This is because one of the main advantages of skforecast is that it helps the user to create training matrices for autoregressive forecasting.

If you are trying to generate the predictions for a time series to fill in all the missing values, but without first filling in the data, what I can recommend is to use weights in your prediction to exclude these observations.

1. Input missing values to avoid NaNs.
2. Train the model and give 0 weight to these observations.
3. Get predictions on training data (you can use in-sample predictions or backtesting).

Take a look at this example of forecasting with missing values: https://skforecast.org/latest/faq/forecasting-time-series-with-missing-values#forecasting-time-series-with-missing-values

Predictions on training data:

https://skforecast.org/latest/user_guides/autoregresive-forecaster#prediction-on-training-data

Using backtesting: https://skforecast.org/latest/user_guides/backtesting#backtesting-on-training-data

Hope it helps!

**Reply by slavakx** [reply]
Hi @JavierEscobarOrtiz 

Thanks for these tips! I have some follow-up questions regarding the approach:

* Does this approach work also together with exogenous variables?
* Would it work when I want to model all available time series in one model? 
* What would be the strategy for missing value imputation along a very long time period? Let's say training a model with 7 day ahead forecast, update the historical timeseries with imputed values, re-training the model again for the next 7 days? 

Thanks!

**Reply by JavierEscobarOrtiz** [reply]
> Does this approach work also together with exogenous variables?

Yes, when you assign a weight, it affects all the columns (lags + exogenous variables) corresponding to that index.

> Would it work when I want to model all available time series in one model?

Yes, but keep in mind that the weights in the ForecasterMultiSeries require additional configuration (not much). Have a look at: https://skforecast.org/latest/user_guides/independent-multi-time-series-forecasting#weights-in-multi-series

> What would be the strategy for missing value imputation along a very long time period? Let's say training a model with 7 day ahead forecast, update the historical timeseries with imputed values, re-training the model again for the next 7 days?

I will recommend you start with backtesting, you have 2 approaches here:

1. Train your model with your whole dataset (inputting missing observations and assigning them 0 weight) and then do a [backtesting on the training data](https://skforecast.org/latest/user_guides/backtesting#backtesting-on-training-data). Then I merge your original data with the training predictions, filling in the NaNs with those predictions.

2. Use [backtesting with refit](https://skforecast.org/latest/user_guides/backtesting#backtest) to generate predictions and then merge them with your original data. In this approach, the filled values are predictions generated with unseen data.

![Backtesting refit](https://skforecast.org/0.12.1/img/backtesting_refit.gif)

I recommend that you try both approaches and see which one fits your use case better. 😄

**Reply by slavakx** [reply]
Wow, thanks for your input, @JavierEscobarOrtiz !
I am going to try both approaches. 

Could you provide a description of how the weighting mechanism works? Is it implemented independently, or does it leverage the capabilities of the underlying estimator, such as LightGBM?

**Reply by JavierEscobarOrtiz** [reply]
❤️ 

As you said, we use the capabilities of the underlying estimator with the common fit argument `sample_weight`.

Under the hood, skforecast just generates the weight array with the user function (`weight_func`) and aligns it with the training matrix.
